Replace std::isnan/isinf/isfinite UB overloads for time classes with member functions

squelart opened the pull request:
#### acfb618b4d5ad0bc49791190588cfa125d198c6c
<pre>
Replace std::isnan/isinf/isfinite UB overloads for time classes with member functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=264241">https://bugs.webkit.org/show_bug.cgi?id=264241</a>
<a href="https://rdar.apple.com/problem/117986797">rdar://problem/117986797</a>

Reviewed by Dan Glastonbury.

Time types under WTF (MonotonicTime, etc.) add overloads of std::isnan, isinf, and isfinite,
but it is generally UB to add declarations&amp;definitions to `namespace std`:
<a href="https://en.cppreference.com/w/cpp/language/extending_std">https://en.cppreference.com/w/cpp/language/extending_std</a>

Instead, these functions should be class member functions, or free functions outside of `std`.
I don&apos;t believe these functions are used in generic code, so I think member functions are more
appropriate here.

* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::didGarbageCollect):
* Source/WTF/wtf/ApproximateTime.cpp:
(WTF::ApproximateTime::approximateWallTime const):
(WTF::ApproximateTime::approximateMonotonicTime const):
* Source/WTF/wtf/ApproximateTime.h:
(WTF::ApproximateTime::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/GenericTimeMixin.h:
(WTF::GenericTimeMixin::isNaN const):
(WTF::GenericTimeMixin::isInfinity const):
(WTF::GenericTimeMixin::isFinite const):
(WTF::GenericTimeMixin::timePointFromNow):
* Source/WTF/wtf/MonotonicTime.cpp:
(WTF::MonotonicTime::approximateWallTime const):
* Source/WTF/wtf/MonotonicTime.h:
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/Seconds.h:
(WTF::Seconds::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/Stopwatch.h:
(WTF::Stopwatch::start):
(WTF::Stopwatch::stop):
(WTF::Stopwatch::fromMonotonicTime const):
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::hasElapsed):
* Source/WTF/wtf/TimeWithDynamicClockType.h:
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/WallTime.cpp:
(WTF::WallTime::approximateMonotonicTime const):
* Source/WTF/wtf/WallTime.h:
(WTF::WallTime::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/WorkerPool.cpp:
(WTF::WorkerPool::shouldSleep):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::ThreadCondition::timedWait):
* Source/WTF/wtf/win/ThreadingWin.cpp:
(WTF::absoluteTimeToWaitTimeoutInterval):
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePayDeferredPaymentRequest::validate const):
* Source/WebCore/Modules/applepay/cocoa/PaymentSummaryItemsCocoa.mm:
(WebCore::platformRecurringSummaryItem):
(WebCore::platformDeferredSummaryItem):
* Source/WebCore/bindings/IDLTypes.h:
(WebCore::IDLDate::isNullValue):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::createIDBKeyFromValue):
* Source/WebCore/fileapi/FileReader.cpp:
(WebCore::FileReader::didReceiveData):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::finalizeFrame):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::consumeLastActivationIfNecessary):
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::gcTimerString):
* Source/WebCore/page/linux/ResourceUsageOverlayLinux.cpp:
(WebCore::gcTimerString):
* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
(WebCore::TimerBase::nextUnalignedFireInterval const):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(CachedResourceStreamingClient::dataReceived):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::elapsedTime):
* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::tick):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::elapsedTime):
* Source/WebKit/Platform/IPC/Timeout.h:
(IPC::Timeout::Timeout):
(IPC::Timeout::isInfinity const):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI):
* Source/WebKit/Shared/ApplePay/cocoa/DeferredPaymentRequestCocoa.mm:
(WebKit::platformDeferredPaymentRequest):
* Tools/TestWebKitAPI/Tests/WTF/Time.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270366@main">https://commits.webkit.org/270366@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4826ff6c830e34fb7461439c9cbc21644fdcf5e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25087 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27202 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23026 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5332 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1064 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23290 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21655 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27781 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2354 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22591 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28722 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21817 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22892 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22947 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26534 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24332 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2295 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/603 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31742 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3649 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6952 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6059 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2741 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2638 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
🚢 r+ (not reviewer yet)

djg reviewed (approved):
> Instead, these functions should be class member functions, or free functions outside of `std`.
I don't believe these functions are used in generic code, so I think member functions are more
appropriate here.

Personally, I would always prefer a free function over a method, unless it really has to be a method, but that's not the C++ style.

What's here is not wrong, so :shipit:, unless others have stronger opinions.

squelart committed:
Replace std::isnan/isinf/isfinite UB overloads for time classes with member functions
https://bugs.webkit.org/show_bug.cgi?id=264241
rdar://problem/117986797

Reviewed by Dan Glastonbury.

Time types under WTF (MonotonicTime, etc.) add overloads of std::isnan, isinf, and isfinite,
but it is generally UB to add declarations&definitions to `namespace std`:
https://en.cppreference.com/w/cpp/language/extending_std

Instead, these functions should be class member functions, or free functions outside of `std`.
I don't believe these functions are used in generic code, so I think member functions are more
appropriate here.

* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::didGarbageCollect):
* Source/WTF/wtf/ApproximateTime.cpp:
(WTF::ApproximateTime::approximateWallTime const):
(WTF::ApproximateTime::approximateMonotonicTime const):
* Source/WTF/wtf/ApproximateTime.h:
(WTF::ApproximateTime::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/GenericTimeMixin.h:
(WTF::GenericTimeMixin::isNaN const):
(WTF::GenericTimeMixin::isInfinity const):
(WTF::GenericTimeMixin::isFinite const):
(WTF::GenericTimeMixin::timePointFromNow):
* Source/WTF/wtf/MonotonicTime.cpp:
(WTF::MonotonicTime::approximateWallTime const):
* Source/WTF/wtf/MonotonicTime.h:
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/Seconds.h:
(WTF::Seconds::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/Stopwatch.h:
(WTF::Stopwatch::start):
(WTF::Stopwatch::stop):
(WTF::Stopwatch::fromMonotonicTime const):
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::hasElapsed):
* Source/WTF/wtf/TimeWithDynamicClockType.h:
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/WallTime.cpp:
(WTF::WallTime::approximateMonotonicTime const):
* Source/WTF/wtf/WallTime.h:
(WTF::WallTime::MarkableTraits::isEmptyValue):
(std::isnan): Deleted.
(std::isinf): Deleted.
(std::isfinite): Deleted.
* Source/WTF/wtf/WorkerPool.cpp:
(WTF::WorkerPool::shouldSleep):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::ThreadCondition::timedWait):
* Source/WTF/wtf/win/ThreadingWin.cpp:
(WTF::absoluteTimeToWaitTimeoutInterval):
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePayDeferredPaymentRequest::validate const):
* Source/WebCore/Modules/applepay/cocoa/PaymentSummaryItemsCocoa.mm:
(WebCore::platformRecurringSummaryItem):
(WebCore::platformDeferredSummaryItem):
* Source/WebCore/bindings/IDLTypes.h:
(WebCore::IDLDate::isNullValue):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::createIDBKeyFromValue):
* Source/WebCore/fileapi/FileReader.cpp:
(WebCore::FileReader::didReceiveData):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::finalizeFrame):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::consumeLastActivationIfNecessary):
* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::gcTimerString):
* Source/WebCore/page/linux/ResourceUsageOverlayLinux.cpp:
(WebCore::gcTimerString):
* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
(WebCore::TimerBase::nextUnalignedFireInterval const):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(CachedResourceStreamingClient::dataReceived):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::elapsedTime):
* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::tick):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::elapsedTime):
* Source/WebKit/Platform/IPC/Timeout.h:
(IPC::Timeout::Timeout):
(IPC::Timeout::isInfinity const):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI):
* Source/WebKit/Shared/ApplePay/cocoa/DeferredPaymentRequestCocoa.mm:
(WebKit::platformDeferredPaymentRequest):
* Tools/TestWebKitAPI/Tests/WTF/Time.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270366@main
